### PR TITLE
Implement AsRef<Path> for Component

### LINKED ISSUE #41866: Collecting into a PathBuf from Iterator<Item=Component> is hard.
See discussion here: https://github.com/rust-lang/rust/pull/41531.

To summarize, the following should work, but currently doesn't. It's likely that implementing `impl<P: AsRef<OsStr>> FromIterator<P> for PathBuf` would solve this, but I haven't tested.

```rust
use std::path::{PathBuf, Path};

fn main() {
    let path = Path::new("a/b/c/d");
    println!("{:?}", path.components().filter(|c| c.as_os_str() != "b").collect::<PathBuf>());
}
```

### PR DESCRIPTION
**Pull request by Diggsey**
Fixes #41866

### CODE REVIEW

File: src/libstd/path.rs [hunk 1]
```diff
@@ -576,6 +576,13 @@ impl<'a> AsRef<OsStr> for Component<'a> {
     }
 }
 
+#[unstable(feature = "path_component_asref", issue = "41866")]
```
Comment (BurntSushi) [inline src/libstd/path.rs]: Trait impls are insta-stable. Please change this to a `stable` attribute.

**Review by BurntSushi — CHANGES_REQUESTED** [review]

**Comment by BurntSushi** [comment]
This seems reasonable to me. @rfcbot fcp merge

**Comment by Diggsey** [comment]
@BurntSushi is that correct? Is there any documentation about stability attributes?

**Comment by kennytm** [comment]
(@BurntSushi The `@rfcbot fcp merge` should be on its own line)

**Comment by BurntSushi** [comment]
@rfcbot fcp merge

**Comment by BurntSushi** [comment]
@Diggsey 

> is that correct? Is there any documentation about stability attributes?

cc @rust-lang/libs

**Comment by sfackler** [comment]
Yeah, that stability attribute looks right.

**Comment by kennytm** [comment]
Ping @aturon, ticky box for you! https://github.com/rust-lang/rust/pull/46985#issuecomment-354289123

**Comment by vitiral** [comment]
I'm a little confused why adding an `AsRef` caused crates to stop compiling. Could there be a brief triage of why that is (and why rust is going forward with breaking changes)?

**Comment by BurntSushi** [comment]
@vitiral It's well covered ground, but I actually don't know of any central resource on the matter. AFAIK, if your use of `AsRef` breaks then it's generally because it is being misused, and in general, it's not feasible to never add new trait impls.

**Comment by BurntSushi** [comment]
(From a technicality perspective, we're empowered to make changes like this via the [API evolution RFC](https://github.com/nox/rust-rfcs/blob/master/text/1105-api-evolution.md), but pragmatism rules the day.)

**Comment by Diggsey** [comment]
Here is the specific relevant section: https://github.com/rust-lang/rfcs/blob/master/text/1105-api-evolution.md#minor-change-implementing-any-non-fundamental-trait

**Pull request merged**